### PR TITLE
Adding a more descriptive TimeoutError message; fixes #1123

### LINKED ISSUE #1123: `nbconvert` Cell timeout errors don't print cell contents
When a notebook cell times out, nbconvert stops with an error, as expected. However, the documentation says that it should print out _the contents of the cell that had the error_, but it does not:
> Conveniently, the source cell causing the error and the original error name and message are also printed.
- https://nbconvert.readthedocs.io/en/latest/execute_api.html#execution-until-first-error

EDIT: As described below, this is because the source cell doesn't _actually_ return an error, rather `nbconvert`'s monitoring errors. This Issue is to request that we change this behavior to still print the cell that timed out, as if the cell had errored. :) Thanks!

I believe this is because of this "exception while handling the exception":
> `During handling of the above exception, another exception occurred:`

A simple notebook that can reproduce this is a python notebook with just a single cell:
```
import time
time.sleep(10)
```

And then running `nbconvert` with a shorter timeout, e.g. `--ExecutePreprocessor.timeout=2`:
```bash
(15:29:05) $ ~/miniconda3/bin/jupyter nbconvert --to notebook --execute SleepTimeout-10.ipynb --ExecutePreprocessor.timeout=2
[NbConvertApp] WARNING | Config option `template_path` not recognized by `NotebookExporter`.
[NbConvertApp] Converting notebook SleepTimeout-10.ipynb to notebook
[NbConvertApp] Executing notebook with kernel: python3
[NbConvertApp] ERROR | Timeout waiting for execute reply (2s).
Traceback (most recent call last):
  File "/Users/nathan.daly/miniconda3/bin/jupyter-nbconvert", line 11, in <module>
    sys.exit(main())
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/jupyter_core/application.py", line 268, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/traitlets/config/application.py", line 664, in launch_instance
    app.start()
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/nbconvertapp.py", line 340, in start
    self.convert_notebooks()
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/nbconvertapp.py", line 510, in convert_notebooks
    self.convert_single_notebook(notebook_filename)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/nbconvertapp.py", line 481, in convert_single_notebook
    output, resources = self.export_single_notebook(notebook_filename, resources, input_buffer=input_buffer)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/nbconvertapp.py", line 410, in export_single_notebook
    output, resources = self.exporter.from_filename(notebook_filename, resources=resources)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/exporters/exporter.py", line 179, in from_filename
    return self.from_file(f, resources=resources, **kw)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/exporters/exporter.py", line 197, in from_file
    return self.from_notebook_node(nbformat.read(file_stream, as_version=4), resources=resources, **kw)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/exporters/notebook.py", line 32, in from_notebook_node
    nb_copy, resources = super(NotebookExporter, self).from_notebook_node(nb, resources, **kw)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/exporters/exporter.py", line 139, in from_notebook_node
    nb_copy, resources = self._preprocess(nb_copy, resources)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/exporters/exporter.py", line 316, in _preprocess
    nbc, resc = preprocessor(nbc, resc)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/preprocessors/base.py", line 47, in __call__
    return self.preprocess(nb, resources)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/preprocessors/execute.py", line 405, in preprocess
    nb, resources = super(ExecutePreprocessor, self).preprocess(nb, resources)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/preprocessors/base.py", line 69, in preprocess
    nb.cells[index], resources = self.preprocess_cell(cell, resources, index)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/preprocessors/execute.py", line 438, in preprocess_cell
    reply, outputs = self.run_cell(cell, cell_index, store_history)
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/preprocessors/execute.py", line 571, in run_cell
    if self._passed_deadline(deadline):
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/preprocessors/execute.py", line 541, in _passed_deadline
    self._handle_timeout()
  File "/Users/nathan.daly/miniconda3/lib/python3.7/site-packages/nbconvert/preprocessors/execute.py", line 504, in _handle_timeout
    raise TimeoutError("Cell execution timed out")
TimeoutError: Cell execution timed out
```

EDIT: Updated the above output after updating to nbconvert 5.6.0. The original output can be found in a comment below.

Such a notebook is attached here:
[SleepTimeout-10.ipynb.txt](https://github.com/jupyter/nbconvert/files/3757464/SleepTimeout-10.ipynb.txt)

**Nbconvert version:** 5.6.0
```
$ ~/miniconda3/bin/jupyter nbconvert --version
5.6.0
```

---------------------------

Is there anything that can be done so that the culprit cell is still printed out? Why are there two exceptions occurring, the second occurring while handling the first?

### PR DESCRIPTION
**Pull request by jonbannister**

### CODE REVIEW

File: nbconvert/preprocessors/execute.py [hunk 1]
```diff
@@ -494,14 +517,14 @@ def _get_timeout(self, cell):
 
         return timeout
 
-    def _handle_timeout(self):
+    def _handle_timeout(self, timeout, cell=None):
         self.log.error(
-            "Timeout waiting for execute reply (%is)." % self.timeout)
+            "Timeout waiting for execute reply (%is)." % timeout)
```
Comment (jonbannister) [inline nbconvert/preprocessors/execute.py]: This was because the specified `timeout` is not always `self.timeout`, i.e. in the case of the IOPub timeout.

**Comment by takluyver** [comment]
Unfortunately we've now got 2 pull requests addressing the same issue (this and #1129). Sorry, that's my mistake - I wasn't keeping track closely enough of who was tackling what.

Still, trying to turn a negative into a positive: that puts you and @joy-rosie in a good position to review each other's pull requests - you've both understood and addressed the problem, so you can evaluate what the other PR does. Ultimately, we can probably only merge one, but you're still contributing to the project by reviewing a PR.

**Comment by jonbannister** [comment]
Argh, what a disaster! I've had a look at the other PR and have applied the things I think I missed (i.e. defending against a `cell=None`) and left some comments. Happy to withdraw this one as I submitted last, though!

**Comment by joy-rosie** [comment]
> I do think this PR is a little more ready to merge than the other. If @joy-rosie wanted to pull change they made into this PR as review or a commit I think that'd be the cleanest / easier to merge.

I am a bit new to this, would the workflow be that I make a fork on @jonbannister branch then pull my changes in and finally then pull the changes into this pull request?

Because I cannot seem to be able do a direct pull request from my fork of jupyter master to this.

**Comment by MSeal** [comment]
There a few approaches here to get changes copied over. You could checkout this PR branch, https://github.com/jonbannister/nbconvert/tree/more-descriptive-timeout, and add your change to to form a new branch. Or you could copy the parts over to your branch by either doing a merge against the above branch if you think the changes are close enough to resolve the conflicting changes, or you can do it yourself by selecting the code parts that seem sharable and reusable in an editor. You can also do the opposite, which is what I think you were trying, which is to git checkout the `more-descriptive-timeout` branch and then doing a git merge `cell_timeout_issue_1123`.

**Comment by jonbannister** [comment]
Shall we merge this? I believe the docs fail is unrelated.

**Comment by MSeal** [comment]
Yes. I'm going to merge for now and folks can make additional PRs if they want to pull additional changes later.

**Pull request merged**